File: internal/erdesigner/join_pathfinder.go
// Copyright (c) 2026 Technarion Oy. All rights reserved.
//
// This software and its source code are proprietary and confidential.
// Unauthorized copying, distribution, modification, or use of this software,
// in whole or in part, is strictly prohibited without prior written permission
// from Technarion Oy.
//
// Commercial use of this software is restricted to parties holding a valid
// license agreement with Technarion Oy.

package erdesigner

import (
	"fmt"
	"strings"

	"thaw/internal/snowflake"
)

// ── Types (JSON-serializable for Wails IPC) ─────────────────────────────────

// TableRef identifies a table by schema and name.
type TableRef struct {
	Schema string `json:"schema"`
	Name   string `json:"name"`
}

// FKColRef identifies a column in a specific table.
type FKColRef struct {
	Schema string `json:"schema"`
	Table  string `json:"table"`
	Col    string `json:"col"`
}

// FKPair represents a single column-to-column FK mapping within a join.
type FKPair struct {
	From FKColRef `json:"from"`
	To   FKColRef `json:"to"`
}

// JoinPathEdge is an edge in a join path carrying FK column references.
type JoinPathEdge struct {
	From FKColRef `json:"from"`
	To   FKColRef `json:"to"`
}

// JoinPath is a sequence of tables and edges connecting them via FK
// relationships, as returned by FindJoinPaths.
type JoinPath struct {
	Tables []TableRef     `json:"tables"`
	Edges  []JoinPathEdge `json:"edges"`
}

// JoinEntry represents a single JOIN clause in the built query state.
type JoinEntry struct {
	Table          TableRef `json:"table"`
	JoinType       string   `json:"joinType"`
	OnCondition    string   `json:"onCondition"`
	FKPairs        []FKPair `json:"fkPairs"`
	IsIntermediate bool     `json:"isIntermediate"`
}

// JoinQueryState holds the complete state for generating a JOIN SQL query.
type JoinQueryState struct {
	Database        string              `json:"database"`
	BaseTable       TableRef            `json:"baseTable"`
	Joins           []JoinEntry         `json:"joins"`
	SelectedColumns map[string][]string `json:"selectedColumns"`
}

// ── Internal adjacency types ────────────────────────────────────────────────

// adjEdge is an edge in the adjacency list, carrying FK column info.
type adjEdge struct {
	neighbor                                        string
	fromSchema, fromTable, fromCol                  string
	toSchema, toTable, toCol                        string
}

// bfsResult holds a single BFS path and its traversed edges.
type bfsResult struct {
	path  []string
	edges []adjEdge
}

// ── Adjacency graph ─────────────────────────────────────────────────────────

// buildAdjacency builds a bidirectional adjacency list from FK edges.
func buildAdjacency(fks []snowflake.ERForeignKey) map[string][]adjEdge {
	adj := make(map[string][]adjEdge)
	for _, fk := range fks {
		fromKey := snowflake.TableKey(fk.FromSchema, fk.FromTable)
		toKey := snowflake.TableKey(fk.ToSchema, fk.ToTable)

		// Forward: from -> to
		adj[fromKey] = append(adj[fromKey], adjEdge{
			neighbor:   toKey,
			fromSchema: fk.FromSchema, fromTable: fk.FromTable, fromCol: fk.FromCol,
			toSchema: fk.ToSchema, toTable: fk.ToTable, toCol: fk.ToCol,
		})

		// Reverse: to -> from (same FK info, traversed backwards)
		adj[toKey] = append(adj[toKey], adjEdge{
			neighbor:   fromKey,
			fromSchema: fk.FromSchema, fromTable: fk.FromTable, fromCol: fk.FromCol,
			toSchema: fk.ToSchema, toTable: fk.ToTable, toCol: fk.ToCol,
		})
	}
	return adj
}

// ── BFS ─────────────────────────────────────────────────────────────────────

type parentEntry struct {
	parent string
	edge   adjEdge
}

// bfsAllShortest returns all shortest paths from start to end, capped at
// maxPaths to avoid overwhelming the disambiguation UI in dense schemas.
func bfsAllShortest(adj map[string][]adjEdge, start, end string, maxPaths int) []bfsResult {
	if start == end {
		return []bfsResult{{path: []string{start}}}
	}

	parents := make(map[string][]parentEntry)
	dist := make(map[string]int)
	dist[start] = 0
	parents[start] = nil

	queue := []string{start}
	head := 0
	found := false
	targetDist := int(^uint(0) >> 1) // max int

	for head < len(queue) {
		current := queue[head]
		head++
		currentDist := dist[current]

		if found && currentDist >= targetDist {
			break
		}

		for _, edge := range adj[current] {
			next := edge.neighbor
			nextDist := currentDist + 1

			if _, seen := dist[next]; !seen {
				dist[next] = nextDist
				parents[next] = []parentEntry{{parent: current, edge: edge}}
				queue = append(queue, next)

				if next == end {
					found = true
					targetDist = nextDist
				}
			} else if dist[next] == nextDist {
				parents[next] = append(parents[next], parentEntry{parent: current, edge: edge})
			}
		}
	}

	if !found {
		return nil
	}

	// Reconstruct shortest paths via backtracking.
	var results []bfsResult
	var backtrack func(node string, path []string, edges []adjEdge)
	backtrack = func(node string, path []string, edges []adjEdge) {
		if len(results) >= maxPaths {
			return
		}
		if node == start {
			// Reverse copies of path and edges.
			rp := make([]string, len(path))
			re := make([]adjEdge, len(edges))
			for i, j := 0, len(path)-1; i < len(path); i, j = i+1, j-1 {
				rp[i] = path[j]
			}
			for i, j := 0, len(edges)-1; i < len(edges); i, j = i+1, j-1 {
				re[i] = edges[j]
			}
			results = append(results, bfsResult{path: rp, edges: re})
			return
		}
		for _, pe := range parents[node] {
			if len(results) >= maxPaths {
				return
			}
			backtrack(pe.parent, append(path, pe.parent), append(edges, pe.edge))
		}
	}

	backtrack(end, []string{end}, nil)
	return results
}

// parseKey splits a table key "SCHEMA.TABLE" back into schema + name.
func parseKey(key string) TableRef {
	schema, name, _ := strings.Cut(key, ".")
	return TableRef{Schema: schema, Name: name}
}

// bfsResultToJoinPath converts a bfsResult to a JoinPath.
func bfsResultToJoinPath(r bfsResult) JoinPath {
	tables := make([]TableRef, len(r.path))
	for i, k := range r.path {
		tables[i] = parseKey(k)
	}
	edges := make([]JoinPathEdge, len(r.edges))
	for i, e := range r.edges {
		edges[i] = JoinPathEdge{
			From: FKColRef{Schema: e.fromSchema, Table: e.fromTable, Col: e.fromCol},
			To:   FKColRef{Schema: e.toSchema, Table: e.toTable, Col: e.toCol},
		}
	}
	return JoinPath{Tables: tables, Edges: edges}
}

// ── Public API ──────────────────────────────────────────────────────────────

// FindJoinPaths finds join paths connecting the selected tables via FK
// relationships.
//
// For 2 tables: BFS from A to B, returns all shortest paths (up to 10).
// For 3+ tables: Steiner tree approximation — iteratively connects each
// unconnected table to the nearest already-connected table via BFS.
//
// Returns an empty slice if tables cannot be connected or fewer than 2 are
// selected.
func FindJoinPaths(selectedTables []TableRef, fks []snowflake.ERForeignKey) []JoinPath {
	if len(selectedTables) < 2 {
		return []JoinPath{}
	}

	adj := buildAdjacency(fks)
	selectedKeys := make([]string, len(selectedTables))
	for i, t := range selectedTables {
		selectedKeys[i] = snowflake.TableKey(t.Schema, t.Name)
	}

	// Check all selected tables exist in the graph.
	for _, key := range selectedKeys {
		if _, ok := adj[key]; !ok {
			return []JoinPath{}
		}
	}

	if len(selectedKeys) == 2 {
		results := bfsAllShortest(adj, selectedKeys[0], selectedKeys[1], 10)
		paths := make([]JoinPath, len(results))
		for i, r := range results {
			paths[i] = bfsResultToJoinPath(r)
		}
		return paths
	}

	// 3+ tables: Steiner tree approximation (greedy).
	connected := map[string]bool{selectedKeys[0]: true}
	treeKeys := []string{selectedKeys[0]}
	treeKeySet := map[string]bool{selectedKeys[0]: true}
	var treeEdges []adjEdge
	treeEdgeKeys := map[string]bool{}
	remaining := map[string]bool{}
	for _, k := range selectedKeys[1:] {
		remaining[k] = true
	}

	for len(remaining) > 0 {
		var bestPath *bfsResult
		var bestTarget string

		for target := range remaining {
			for source := range connected {
				paths := bfsAllShortest(adj, source, target, 1)
				if len(paths) == 0 {
					continue
				}
				if bestPath == nil || len(paths[0].path) < len(bestPath.path) {
					p := paths[0]
					bestPath = &p
					bestTarget = target
				}
			}
		}

		if bestPath == nil {
			return []JoinPath{} // disconnected graph
		}

		for _, key := range bestPath.path {
			connected[key] = true
			if !treeKeySet[key] {
				treeKeySet[key] = true
				treeKeys = append(treeKeys, key)
			}
		}
		for _, e := range bestPath.edges {
			eKey := fmt.Sprintf("%s.%s.%s-%s.%s.%s",
				e.fromSchema, e.fromTable, e.fromCol,
				e.toSchema, e.toTable, e.toCol)
			if !treeEdgeKeys[eKey] {
				treeEdgeKeys[eKey] = true
				treeEdges = append(treeEdges, e)
			}
		}
		delete(remaining, bestTarget)
	}

	edges := make([]JoinPathEdge, len(treeEdges))
	for i, e := range treeEdges {
		edges[i] = JoinPathEdge{
			From: FKColRef{Schema: e.fromSchema, Table: e.fromTable, Col: e.fromCol},
			To:   FKColRef{Schema: e.toSchema, Table: e.toTable, Col: e.toCol},
		}
	}
	tables := make([]TableRef, len(treeKeys))
	for i, k := range treeKeys {
		tables[i] = parseKey(k)
	}

	return []JoinPath{{Tables: tables, Edges: edges}}
}

// BuildJoinState converts a JoinPath into a JoinQueryState ready for SQL
// generation. The first table in the path becomes the base table. Tables
// not in the user's original selection are marked as intermediate.
func BuildJoinState(path JoinPath, selectedTables []TableRef, database string) JoinQueryState {
	selectedSet := make(map[string]bool, len(selectedTables))
	for _, t := range selectedTables {
		selectedSet[snowflake.TableKey(t.Schema, t.Name)] = true
	}

	baseTable := path.Tables[0]
	visited := map[string]bool{snowflake.TableKey(baseTable.Schema, baseTable.Name): true}
	joins := make([]JoinEntry, 0, len(path.Tables)-1)

	for i := 1; i < len(path.Tables); i++ {
		table := path.Tables[i]
		tKey := snowflake.TableKey(table.Schema, table.Name)

		var conditions []string
		var fkPairs []FKPair
		for _, edge := range path.Edges {
			fromKey := snowflake.TableKey(edge.From.Schema, edge.From.Table)
			toKey := snowflake.TableKey(edge.To.Schema, edge.To.Table)

			if (fromKey == tKey && visited[toKey]) || (toKey == tKey && visited[fromKey]) {
				conditions = append(conditions, fmt.Sprintf(
					"%s.%s.%s = %s.%s.%s",
					edge.From.Schema, edge.From.Table, edge.From.Col,
					edge.To.Schema, edge.To.Table, edge.To.Col,
				))
				fkPairs = append(fkPairs, FKPair{
					From: FKColRef{Schema: edge.From.Schema, Table: edge.From.Table, Col: edge.From.Col},
					To:   FKColRef{Schema: edge.To.Schema, Table: edge.To.Table, Col: edge.To.Col},
				})
			}
		}

		visited[tKey] = true

		onCondition := strings.Join(conditions, " AND ")

		joins = append(joins, JoinEntry{
			Table:          TableRef{Schema: table.Schema, Name: table.Name},
			JoinType:       "INNER",
			OnCondition:    onCondition,
			FKPairs:        fkPairs,
			IsIntermediate: !selectedSet[tKey],
		})
	}

	return JoinQueryState{
		Database:        database,
		BaseTable:       TableRef{Schema: baseTable.Schema, Name: baseTable.Name},
		Joins:           joins,
		SelectedColumns: map[string][]string{},
	}
}

File: frontend/src/components/er/joinPathfinder.ts
// Copyright (c) 2026 Technarion Oy. All rights reserved.
// @thaw-domain: ER Designer

import type { JoinPath, JoinQueryState, JoinEntry } from "./erTypes";

/** FK edge in the adjacency graph. */
interface FKEdge {
  fromSchema: string;
  fromTable: string;
  fromCol: string;
  toSchema: string;
  toTable: string;
  toCol: string;
}

/** Canonical key for a table node: "SCHEMA.TABLE" (uppercase). */
function tableKey(schema: string, name: string): string {
  return `${schema.toUpperCase()}.${name.trim().toUpperCase()}`;
}

/** Parse a table key back into schema + name. */
function parseKey(key: string): { schema: string; name: string } {
  const dot = key.indexOf(".");
  return { schema: key.slice(0, dot), name: key.slice(dot + 1) };
}

/** An edge in the adjacency list, carrying FK column info and direction. */
interface AdjEdge {
  neighbor: string; // tableKey of the other table
  fromSchema: string;
  fromTable: string;
  fromCol: string;
  toSchema: string;
  toTable: string;
  toCol: string;
}

/**
 * Build a bidirectional adjacency list from FK edges.
 * Both directions are stored so BFS can traverse parent->child and child->parent.
 */
function buildAdjacency(fks: FKEdge[]): Map<string, AdjEdge[]> {
  const adj = new Map<string, AdjEdge[]>();

  const ensure = (key: string) => {
    if (!adj.has(key)) adj.set(key, []);
  };

  for (const fk of fks) {
    const fromKey = tableKey(fk.fromSchema, fk.fromTable);
    const toKey = tableKey(fk.toSchema, fk.toTable);
    ensure(fromKey);
    ensure(toKey);

    // Forward: from -> to
    adj.get(fromKey)!.push({
      neighbor: toKey,
      fromSchema: fk.fromSchema,
      fromTable: fk.fromTable,
      fromCol: fk.fromCol,
      toSchema: fk.toSchema,
      toTable: fk.toTable,
      toCol: fk.toCol,
    });

    // Reverse: to -> from (same FK info, just traversed backwards)
    adj.get(toKey)!.push({
      neighbor: fromKey,
      fromSchema: fk.fromSchema,
      fromTable: fk.fromTable,
      fromCol: fk.fromCol,
      toSchema: fk.toSchema,
      toTable: fk.toTable,
      toCol: fk.toCol,
    });
  }

  return adj;
}

interface BFSResult {
  path: string[];       // table keys from start to end
  edges: AdjEdge[];     // one edge per hop
}

/**
 * BFS from `start` to `end`, returning ALL shortest paths of equal length.
 * Each path is a sequence of table keys + the edges traversed.
 */
function bfsAllShortest(
  adj: Map<string, AdjEdge[]>,
  start: string,
  end: string,
): BFSResult[] {
  if (start === end) return [{ path: [start], edges: [] }];

  // BFS tracking: for each node, store all parent entries (multiple = multiple shortest paths)
  const parents = new Map<string, { parent: string; edge: AdjEdge }[]>();
  const dist = new Map<string, number>();
  dist.set(start, 0);
  parents.set(start, []);

  const queue: string[] = [start];
  let found = false;
  let targetDist = Infinity;

  while (queue.length > 0) {
    const current = queue.shift()!;
    const currentDist = dist.get(current)!;

    if (found && currentDist >= targetDist) break;

    for (const edge of adj.get(current) ?? []) {
      const next = edge.neighbor;
      const nextDist = currentDist + 1;

      if (!dist.has(next)) {
        dist.set(next, nextDist);
        parents.set(next, [{ parent: current, edge }]);
        queue.push(next);

        if (next === end) {
          found = true;
          targetDist = nextDist;
        }
      } else if (dist.get(next) === nextDist) {
        // Same distance — add another parent (alternative shortest path)
        parents.get(next)!.push({ parent: current, edge });
      }
    }
  }

  if (!found) return [];

  // Reconstruct all shortest paths via backtracking
  const results: BFSResult[] = [];

  function backtrack(node: string, path: string[], edges: AdjEdge[]) {
    if (node === start) {
      results.push({ path: [...path].reverse(), edges: [...edges].reverse() });
      return;
    }
    for (const { parent, edge } of parents.get(node) ?? []) {
      path.push(parent);
      edges.push(edge);
      backtrack(parent, path, edges);
      path.pop();
      edges.pop();
    }
  }

  backtrack(end, [end], []);
  return results;
}

/**
 * Find join paths connecting the selected tables via FK relationships.
 *
 * For 2 tables: BFS from A to B, returns all shortest paths.
 * For 3+ tables: Steiner tree approximation — iteratively connects each
 * unconnected table to the nearest already-connected table via BFS.
 *
 * Returns empty array if tables cannot be connected.
 */
export function findJoinPaths(
  selectedTables: { schema: string; name: string }[],
  fks: FKEdge[],
): JoinPath[] {
  if (selectedTables.length < 2) return [];

  const adj = buildAdjacency(fks);
  const selectedKeys = selectedTables.map((t) => tableKey(t.schema, t.name));

  // Check all selected tables exist in the graph
  for (const key of selectedKeys) {
    if (!adj.has(key)) return [];
  }

  if (selectedKeys.length === 2) {
    // Simple case: BFS between two tables, return all shortest paths
    const results = bfsAllShortest(adj, selectedKeys[0], selectedKeys[1]);
    return results.map((r) => bfsResultToJoinPath(r));
  }

  // 3+ tables: Steiner tree approximation
  // Start from first selected table, iteratively BFS to nearest unconnected
  const connected = new Set<string>([selectedKeys[0]]);
  const treeKeys: string[] = [selectedKeys[0]];
  const treeEdges: AdjEdge[] = [];
  const remaining = new Set(selectedKeys.slice(1));

  while (remaining.size > 0) {
    let bestPath: BFSResult | null = null;
    let bestTarget: string | null = null;

    for (const target of remaining) {
      // BFS from each connected node to the target, keep shortest
      for (const source of connected) {
        const paths = bfsAllShortest(adj, source, target);
        if (paths.length === 0) continue;
        // Take first shortest path
        if (!bestPath || paths[0].path.length < bestPath.path.length) {
          bestPath = paths[0];
          bestTarget = target;
        }
      }
    }

    if (!bestPath || !bestTarget) return []; // Disconnected graph

    // Add all nodes and edges from this path to the tree
    for (const key of bestPath.path) {
      connected.add(key);
      if (!treeKeys.includes(key)) treeKeys.push(key);
    }
    treeEdges.push(...bestPath.edges);
    remaining.delete(bestTarget);
  }

  // Build a single JoinPath from the Steiner tree
  const joinPath: JoinPath = {
    tables: treeKeys.map(parseKey),
    edges: treeEdges.map((e) => ({
      from: { schema: e.fromSchema, table: e.fromTable, col: e.fromCol },
      to: { schema: e.toSchema, table: e.toTable, col: e.toCol },
    })),
  };

  return [joinPath];
}

function bfsResultToJoinPath(result: BFSResult): JoinPath {
  return {
    tables: result.path.map(parseKey),
    edges: result.edges.map((e) => ({
      from: { schema: e.fromSchema, table: e.fromTable, col: e.fromCol },
      to: { schema: e.toSchema, table: e.toTable, col: e.toCol },
    })),
  };
}

/**
 * Convert a JoinPath into a JoinQueryState ready for SQL generation.
 * The first table in the path becomes the base table.
 * Tables not in the user's original selection are marked as intermediate.
 */
export function buildJoinState(
  path: JoinPath,
  selectedTables: { schema: string; name: string }[],
  _fks: FKEdge[],
): JoinQueryState {
  const selectedSet = new Set(
    selectedTables.map((t) => tableKey(t.schema, t.name)),
  );

  const baseTable = path.tables[0];
  const joins: JoinEntry[] = [];
  const visited = new Set<string>([tableKey(baseTable.schema, baseTable.name)]);

  // Build ON conditions from edges, grouping by table pair for composite FKs
  // Walk the path tables in order, finding the edges that connect each to the tree
  for (let i = 1; i < path.tables.length; i++) {
    const table = path.tables[i];
    const tKey = tableKey(table.schema, table.name);

    // Find all edges connecting this table to any already-visited table
    const conditions: string[] = [];
    for (const edge of path.edges) {
      const fromKey = tableKey(edge.from.schema, edge.from.table);
      const toKey = tableKey(edge.to.schema, edge.to.table);

      if (
        (fromKey === tKey && visited.has(toKey)) ||
        (toKey === tKey && visited.has(fromKey))
      ) {
        conditions.push(
          `${edge.from.schema}.${edge.from.table}.${edge.from.col} = ${edge.to.schema}.${edge.to.table}.${edge.to.col}`,
        );
      }
    }

    visited.add(tKey);

    joins.push({
      table: { schema: table.schema, name: table.name },
      joinType: "INNER",
      onCondition: conditions.join(" AND "),
      isIntermediate: !selectedSet.has(tKey),
    });
  }

  return {
    baseTable: { schema: baseTable.schema, name: baseTable.name },
    joins,
    selectedColumns: new Map(),
  };
}
